### PR TITLE
Add CR to panic backtrace

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

**Comment by bjorn3** [comment]
Example of current behavior:
```
abrr>vreiooieoeriotothread 'main' panicked at 'cannot remove a char from the end of a string', /checkout/src/liballoc/string.rs:1020:20
                                                       note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
                                                              stack backtrace:
                                                                                 0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
                                                                         at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
                                                          1: std::sys_common::backtrace::_print
                            at /checkout/src/libstd/sys_common/backtrace.rs:71
                                                                                 2: std::panicking::default_hook::{{closure}}
                                                          at /checkout/src/libstd/sys_common/backtrace.rs:60
                                         at /checkout/src/libstd/panicking.rs:380
    3: std::panicking::default_hook
                                                at /checkout/src/libstd/panicking.rs:396
           4: std::panicking::rust_panic_with_hook
                                                               at /checkout/src/libstd/panicking.rs:611
                          5: std::panicking::begin_panic_new
                                                                         at /checkout/src/libstd/panicking.rs:553
                                    6: std::panicking::begin_panic_fmt
                                                                                   at /checkout/src/libstd/panicking.rs:521
                                              7: rust_begin_unwind
                                                                               at /checkout/src/libstd/panicking.rs:497
                                          8: core::panicking::panic_fmt
                                                                                    at /checkout/src/libcore/panicking.rs:92
                                               9: core::panicking::panic
                                                                                     at /checkout/src/libcore/panicking.rs:51
                                               10: alloc::string::String::remove
                                                                                            at /checkout/obj/<panic macros>:4
                                               11: mouse_readline::read_line
                                                                                         at src/lib.rs:24
                           12: read::main
                                                      at examples/read.rs:4
```

**Comment by bjorn3** [comment]
Travis is green

**Comment by steveklabnik** [comment]
Hm, where are you seeing this kind of output?

**Comment by bjorn3** [comment]
When panicking while the terminal is in raw mode.

File: src/libstd/panicking.rs [hunk 1]
```diff
@@ -368,7 +368,8 @@ fn default_hook(info: &PanicInfo) {
     let name = thread.as_ref().and_then(|t| t.name()).unwrap_or("<unnamed>");
 
     let write = |err: &mut ::io::Write| {
-        let _ = writeln!(err, "thread '{}' panicked at '{}', {}:{}:{}",
+        // Add a cariage return to reset cursor to the beginning of the line with rawmode tty's
+        let _ = write!(err, "thread '{}' panicked at '{}', {}:{}:{}\r\n",
```
Comment (ranma42) [inline src/libstd/panicking.rs]: Would it be better to do `"thread '{}' panicked at '{}', {}:{}:{}\r\n"` instead?

Comment (bjorn3) [reply src/libstd/panicking.rs]: Isn't that what it is now?

Comment (ranma42) [reply src/libstd/panicking.rs]: oops, my bad, wrong copy & paste. I meant `"\r\nthread '{}' panicked at '{}', {}:{}:{}"`

**Comment by retep998** [comment]
This... doesn't have anything to do with Windows? The paths in the example panic backtrace are clearly unix style paths.

Carriage returns have meaning in terminals. They're not exclusive to `\r\n` newlines on Windows.

**Comment by steveklabnik** [comment]
Yes, i am equally confused.

**Comment by retep998** [comment]
Basically, terminals can be set to raw mode. In raw mode `\n` merely advances you to the next row, without resetting your column back to 0, so you have to explicitly `\r` so you return to the beginning of the line.

**Comment by shepmaster** [comment]
If you have put the terminal in some sort of color mode, then the backtrace will also be in that color, but this doesn't address it. I seem to remember that there are all sorts of modes that a terminal can get into that we don't handle; what is special about raw mode?

**Comment by bjorn3** [comment]
Raw mode makes it very hard to read the backtrace. When coloring the terminal, the fg color is often distinct from the bg color, so that you can still read the backtrace easily.

**Comment by shepmaster** [comment]
Lets say this is a libs issue and pick from that bucket...

r? @dtolnay

**Review by dtolnay — CHANGES_REQUESTED** [review]
I don't think we need to support this. There should be no expectation that output while in raw mode is well behaved.

```
$ cat test.cc
#include <signal.h>
int main() {
  raise(SIGSEGV);
}

$ clang++ --version
clang version 6.0.0-svn317351-1~exp1 (trunk)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ clang++ test.cc -o test

$ ./test 
Segmentation fault (core dumped)

$ stty raw

$ ./test 
                     Segmentation fault (core dumped)
```

**Comment by dtolnay** [comment]
Could this be solved in the application by setting a [panic hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) instead? The hook could set the terminal out of raw mode before entering the panic runtime.

**Comment by bjorn3** [comment]
I didnt know panic hooks were stable. Will use them.

**Pull request closed without merging**